File: app/models/OriginalEmail.java
/**
 * Yobi, Project Hosting SW
 *
 * Copyright 2014 NAVER Corp.
 * http://yobi.io
 *
 * @Author Yi EungJun
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package models;

import models.enumeration.ResourceType;
import models.resource.Resource;
import play.data.validation.Constraints;
import play.db.ebean.Model;

import javax.persistence.*;
import java.util.Date;

@Entity
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"resource_type", "resource_id"}))
public class OriginalEmail extends Model {
    public static final Finder<Long, OriginalEmail> finder = new Finder<>(Long.class,
            OriginalEmail.class);

    private static final long serialVersionUID = 9079975193167733297L;

    @Id
    public Long id;

    @Constraints.Required
    @Column(unique=true)
    public String messageId;

    @Constraints.Required
    @Enumerated(EnumType.STRING)
    public ResourceType resourceType;

    @Constraints.Required
    public String resourceId;

    @Constraints.Required
    private Date handledDate;

    public static OriginalEmail findBy(Resource resource) {
        return finder.where()
                .eq("resourceType", resource.getType())
                .eq("resourceId", resource.getId())
                .findUnique();
    }

    public static boolean exists(Resource resource) {
        return findBy(resource) != null;
    }

    public OriginalEmail(String messageId, Resource resource) {
        this.messageId = messageId;
        this.resourceType = resource.getType();
        this.resourceId = resource.getId();
    }

    @Override
    public void save() {
        handledDate = new Date();
        super.save();
    }

    public Date getHandledDate() {
        return handledDate;
    }
}

File: src/main/kotlin/com/github/yonaprojects/yona/domain/mail/OriginalEmail.kt
package com.github.yonaprojects.yona.domain.mail

import com.github.yonaprojects.yona.domain.enumeration.ResourceType
import jakarta.persistence.*
import java.time.Instant

/**
 * yona의 models/OriginalEmail.java 대응.
 * 수신 이메일의 Message-ID를 그 이메일로 만들어진 리소스(이슈/댓글 등)와 연결한다.
 * 두 가지 용도로 쓰인다:
 *  1) 같은 이메일이 중복 수신됐을 때 이미 처리됐는지 확인(messageId로 조회)
 *  2) 답장(In-Reply-To/References)이 어떤 리소스를 가리키는지 찾기(messageId로 역조회)
 */
@Entity
@Table(
    name = "original_email",
    uniqueConstraints = [UniqueConstraint(columnNames = ["resource_type", "resource_id"])]
)
class OriginalEmail(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Column(nullable = false, unique = true)
    var messageId: String = "",

    @Enumerated(EnumType.STRING)
    @Column(name = "resource_type", nullable = false)
    var resourceType: ResourceType = ResourceType.NOT_A_RESOURCE,

    @Column(name = "resource_id", nullable = false)
    var resourceId: String = "",

    var handledDate: Instant? = Instant.now()
)
